src/chunk_constraint.c

```diff
@@ -188,3 +199,90 @@ chunk_constraint_insert_multi(ChunkConstraint *constraints, Size num_constraints
 
 	heap_close(rel, RowExclusiveLock);
 }
+
+static bool
+chunk_constraint_need_on_chunk(Form_pg_constraint conform)
+{
+	if (conform->contype == CONSTRAINT_CHECK)
+	{
+		/*
+		 * check and not null constraints handled by regular inheritance (from
+		 * docs): All check constraints and not-null constraints on a parent
+		 * table are automatically inherited by its children, unless
+		 * explicitly specified otherwise with NO INHERIT clauses. Other types
+		 * of constraints (unique, primary key, and foreign key constraints)
+		 * are not inherited."
+		 */
+		if (conform->connoinherit)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_IO_OPERATION_NOT_SUPPORTED),
+					 errmsg("NO INHERIT option not supported on hypertables: %s", conform->conname.data)
+					 ));
+		}
+		return false;
+	}
+	return true;
+}
+
+static void
+chunk_constraint_create_on_chunk_impl(Hypertable *ht, Chunk *chunk, HeapTuple constraint_htup)
+{
+	Form_pg_constraint pg_constraint = (Form_pg_constraint) GETSTRUCT(constraint_htup);
+
+	if (chunk_constraint_need_on_chunk(pg_constraint))
+	{
+		Oid			hypertable_constraint_oid = HeapTupleGetOid(constraint_htup);
+		Oid			chunk_constraint_oid;
+		CatalogSecurityContext sec_ctx;
+
+		catalog_become_owner(catalog_get(), &sec_ctx);
+		process_utility_set_expect_chunk_modification(true);
+		chunk_constraint_oid = create_chunk_constraint(chunk, hypertable_constraint_oid);
+		process_utility_set_expect_chunk_modification(false);
+		catalog_restore_user(&sec_ctx);
+
+		Assert(OidIsValid(chunk_constraint_oid));
+
+		if (OidIsValid(pg_constraint->conindid) && pg_constraint->contype != CONSTRAINT_FOREIGN)
+		{
+			chunk_index_create_from_constraint(ht->fd.id, hypertable_constraint_oid, chunk->fd.id, chunk_constraint_oid);
+		}
+	}
+}
+
+void
+chunk_constraint_create_all_on_chunk(Hypertable *ht, Chunk *chunk)
+{
+	ScanKeyData skey;
+	Relation	rel;
+	SysScanDesc scan;
+	HeapTuple	htup;
+
+	ScanKeyInit(&skey,
```

Comment:
On a second look, I am wondering if this is systable scan is really necessary? Isn't this stuff cached in the relcache entry via its `TupleDesc`, i.e., `RelationGetDescr(rel)->constr`?

Reply:
There's also `RelationGetFKeyList()` for foreign keys. Might be worth investigating.

Reply:
I think both of these are less clear. The TupleDesc only has check constraints and RelationGetFKeyList() only has FKs and the latter does a systable scan anyway.